flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -435,6 +439,12 @@ public boolean triggerCheckpoint(long timestamp, boolean isPeriodic) {
 			triggerCheckpoint(timestamp, checkpointProperties, null, isPeriodic, false);
 			return true;
 		} catch (CheckpointException e) {
+			try {
+				long latestGeneratedCheckpointId = getCheckpointIdCounter().getAndIncrement();
```

Comment:
I thought about this but can not make sure if we can get the correct latest checkpoint id when allow concurrent checkpointing.

Reply:
What exactly do you mean? Because the access to the id-generator does not happen under the lock? Then how about just rewriting the trigger checkpoing method:
The catch looks not well designed anyways, instread you could already report the exception to the failure manager instead of throwing, or catch and report while still under the lock inside triggerCheckpoint.
BTW, I have one more important additional point, which is the `numUnsuccessfulCheckpointsTriggers` in checkpoint coordinator, which absolutely sounds like something that should now be moved into the failure manager, wdyt?

Reply:
> I thought about this but can not make sure if we can get the correct latest checkpoint id when allow concurrent checkpointing.

Actually, I mean I cannot get a better way to extract the `latestGeneratedCheckpointId ` except using `CheckpointIDCounter`. I think using `pendingCheckpoints` or `recentPendingCheckpoints` to get the latest pending checkpoint may not accurate when allow concurrent checkpointing. 

What's your suggestion to get the value of this variable?

Reply:
If you read my first comment, it was not about that you get the value from that counter, but I wondered why you do `getAndIncrement` instead of just `get`. Rest of my comments still holds though, why not remove the throwing/catching of an exception that we can also directly report to the failure manager?

Reply:
> If you read my first comment, it was not about that you get the value from that counter, but I wondered why you do `getAndIncrement` instead of just `get`.

Actually, I also want to use `get` API, but `CheckpointIDCounter ` does not provide `get` API.

> Rest of my comments still holds though, why not remove the throwing/catching of an exception that we can also directly report to the failure manager? I also think that makes other counts accurate because it obtaining the id can then always happen under the lock.

OK, agree.

Reply:
> Rest of my comments still holds though, why not remove the throwing/catching of an exception that we can also directly report to the failure manager? I also think that makes other counts accurate because it obtaining the id can then always happen under the lock.

This try/catch block:

```
try {
				long latestGeneratedCheckpointId = getCheckpointIdCounter().getAndIncrement();
				failureManager.handleCheckpointException(e, -1 * latestGeneratedCheckpointId);
			} catch (Exception e1) {
				LOG.warn("Get latest generated checkpoint id error : ", e1);
			}
```
is because the `CheckpointIDCounter#getAndIncrement ` throws a checked `Exception`.

I think if it can change to this code snippet then looks better right:

```
public boolean triggerCheckpoint(long timestamp, boolean isPeriodic) {
		try {
			triggerCheckpoint(timestamp, checkpointProperties, null, isPeriodic, false);
			return true;
		} catch (CheckpointException e) {
			long latestGeneratedCheckpointId = getCheckpointIdCounter().get();
			failureManager.handleCheckpointException(e, -1 * latestGeneratedCheckpointId);
			return false;
		}
	}
```

If the reported behavior exists in the real `triggerCheckpoint` method, it would cause more code to report exception and generate new checkpoint id. In addition, there are two locks exist in the method.

What do you think?

Reply:
Do you think we should provide a `CheckpointIDCounter#get` API?

Reply:
You can add the `#get` if really needed. What I meant is removing the `try-catch`around the `triggerCheckpoint` call completely and exchange the `throw`s inside the method by direct calls to the failure handler.

Reply:
> BTW, I have one more important additional point, which is the `numUnsuccessfulCheckpointsTriggers` in checkpoint coordinator, which absolutely sounds like something that should now be moved into the failure manager, wdyt?

After we provided `CheckpointFailureManager`, IMO the `numUnsuccessfulCheckpointsTriggers` is not valuable. Currently, it is incremented in the trigger lock of method `triggerCheckpoint`. The domain of trigger lock is the subset of the trigger phase which been `CheckpointFailureManager` understood. So the counting is not correct. And it is just for logging purpose. So I suggest we could remove it in this PR or in the third step (next step). If the logging is really necessary, we could do it again after we implemented the new counting logic based on checkpoint id sequence. 

What's your opinion?

Reply:
> What I meant is removing the `try-catch(CheckpointException)`around the `triggerCheckpoint` call completely and exchange the `throw`s inside the method by direct calls to the failure handler.

I have accepted all the suggestion, except this one, because when I was refactoring, I found the complexity may exceed our imagination, some obstacle:

1) 	
```
public boolean triggerCheckpoint(long timestamp, boolean isPeriodic) {
```

This method returns a boolean value, so it needs a binary state, when we catch `CheckpointException` we return false and when we get `PendingCheckpoint` we return true. If the real triggerCheckpoint method does not throw `CheckpointException` how two distinguish binary state?

2)The triggerCheckpoint's boolean return value has been used in so many test case to judge whether a checkpoint is successful or not.

3) The `triggerSavepointInternal` method need to catch the `CheckpointException` to return a completed exceptionally for the future object, that means `triggerSavepointInternal` also needs a binary state for `CompletableFuture<CompletedCheckpoint>`

4) I have implemented `CheckpointIDCounter#get`, after providing this function. I think report failure centralized would be better than scattered code, just like we provide a unified `PendingCheckpoint#abort` before.

What do you think?

Reply:
If having `#get` is already enough and we don't need to do this under the lock then there is no need to refactor. Will take a final look.

Reply:
It seems there are still two test failure issues hang the JVM process, but I cannot find a definite description, one of them is successful in my local test. travis : https://travis-ci.org/apache/flink/builds/532612594

Reply:
I think those problems are related to FLINK-12510, which I will start to review soon. Unlikely that this PR caused the problem.